### PR TITLE
Add callbacks to EnumFields for large ranges

### PR DESCRIPTION
**Pull request by X-Cli**
Hey,

While rewriting parts of the ProfinetIO contribution, I spotted a few lines that, I think, consumes a lot of memory and that could be optimized.

https://github.com/secdev/scapy/blob/master/scapy/contrib/pnio.py#L45-L52

The required optimization is to use a callback where tests can be led on ranges instead of filling up a dict with sometimes up to a million identical strings.

I implemented that feature and added tests for it (and for EnumFields in general, as they were lacking).

@guedou told me to split the regression.uts file so that all tests related to fields would be in a separate unit test file. The rationale is that regression.uts is ever growing and is becoming somewhat bloated.

Cheers,
Florian

### CODE REVIEW

File: scapy/fields.py [hunk 1]
```diff
@@ -774,50 +774,72 @@ def i2repr(self, pkt, x):
...
+                x = self.s2i_cb(x)
+            else:
+                x = self.s2i[x]
         return x
+
     def i2repr_one(self, pkt, x):
-        if self not in conf.noenum and not isinstance(x,VolatileValue) and x in self.i2s:
-            return self.i2s[x]
+        if self not in conf.noenum and not isinstance(x,VolatileValue):
+            if self.i2s is None:
```
Comment (p-l-) [inline scapy/fields.py]: I'd prefer:
```Python
try:
    return self.i2s[x]
except KeyError:
    pass
except TypeError:
    return self.i2s_cb(x)
```

By the way, maybe we could imagine that `.i2s_cb` returns `None` when it has no specific name for a value (that would be the equivalent of a missing key in `.i2s`. In that case, the code could be adapted as:

```Python
try:
    return self.i2s[x]
except KeyError:
    pass
except TypeError:
    val = self.i2s_cb(x)
    if val is not None:
        return val
```

File: scapy/fields.py [hunk 2]
```diff
@@ -859,10 +882,12 @@ def __init__(self, name, default, enum):
 
 class XShortEnumField(ShortEnumField):
     def i2repr_one(self, pkt, x):
-        if self not in conf.noenum and not isinstance(x,VolatileValue) and x in self.i2s:
-            return self.i2s[x]
+        y = super(XShortEnumField, self).i2repr_one(pkt, x)
+        if y != repr(x):
```
Comment (p-l-) [inline scapy/fields.py]: I think we could make sure that this test is unnecessary.

**Comment by X-Cli** [comment]
Thank you for your review and suggestions! 
I always struggle with the pythonish "try and recover" approach, as opposed to the "test/ask permission first" approach. I applied the try/except suggestions and I am OK with them. Thank you.

Concerning https://github.com/secdev/scapy/pull/442#discussion_r95310746, I was not pleased/convinced by this test either. The worst part is the assumption about the default representation that is returned when no user-friendly representation is available. Unfortunately, I have no idea how to avoid it. The current PR duplicates the code from _EnumField.i2repr_one, save for the default case (last return line). How does that sound?

**Comment by X-Cli** [comment]
Requested changes have been applied.
Conflits with newly written unit tests are resolved.
Do you see any other changes required before merging becomes possible?

**Pull request merged**